### PR TITLE
PERF: Get rid of unnececssary allocation in VarReader5.cread_fieldnames

### PR DESCRIPTION
**Pull request by alexhenrie**
This pull request is like #12519 but for the C memory allocation functions.

### CODE REVIEW

File: scipy/io/matlab/mio5_utils.pyx [hunk 1]
```diff
@@ -914,7 +914,7 @@ cdef class VarReader5:
             int *n_duplicates
             char **name_ptrs
         n_duplicates = <int *>calloc(n_names, sizeof(int))
-        name_ptrs = <char **>calloc(n_names, sizeof(char *))
+        name_ptrs = <char **>malloc(n_names * sizeof(char *))
```
Comment (peterbell10) [inline scipy/io/matlab/mio5_utils.pyx:917]: This doesn't even need to be allocated at all. You can change it to
```cython
cdef char * names_ptr = names
```
Then remove all other references except line 927 which should be changed to
```cython
if strcmp(n_ptr, names_ptr[j * namelength]) == 0: # the same
```

Comment (alexhenrie) [reply scipy/io/matlab/mio5_utils.pyx]: Great point! I have made the changes you suggested.

**Comment by charris** [comment]
On Linux calloc is almost free until you write to it.
```
In [4]: %timeit numpy.zeros(10**8)                                              
7.28 µs ± 6.05 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [5]: %timeit numpy.ones(10**8)                                               
68.2 ms ± 137 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
The OS uses virtual pages that are pre-zeroed.

File: scipy/io/matlab/mio5_utils.pyx [hunk 2]
```diff
@@ -914,17 +914,16 @@ cdef class VarReader5:
             int *n_duplicates
             char **name_ptrs
```
Comment (peterbell10) [inline scipy/io/matlab/mio5_utils.pyx:915]: You still need to remove this and the call to `free(name_ptrs)` on line 935.

Comment (alexhenrie) [reply scipy/io/matlab/mio5_utils.pyx]: Done. Sorry I missed that earlier.

**Review by peterbell10 — APPROVED** [review]

**Comment by tylerjereddy** [comment]
CI is green and core dev approval--the change may not provide much performance improvement (as Chuck notes) but it does seem to simplify the code a bit in any case.

Thanks @alexhenrie & reviewers

**Pull request merged**